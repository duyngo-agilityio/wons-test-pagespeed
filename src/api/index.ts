export * from './invoice';
export * from './product';
export * from './statistic';
export * from './user';
export * from './customer';
export * from './task';
export * from './calendar';
export * from './image';
