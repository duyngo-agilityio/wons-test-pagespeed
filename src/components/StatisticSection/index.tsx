// icons
import { FaHeart, FaShoppingBag } from 'react-icons/fa';
import { IoGameControllerSharp } from 'react-icons/io5';
import { MdWork } from 'react-icons/md';

// components
import { StatisticCard } from '@/components';

// models
import { IStatistic } from '@/models';

const StatisticSection = () => {
  const statisticData: IStatistic[] = [
    {
      value: '178+',
      label: 'Favorite Services',
      icon: <FaHeart className="h-6 w-6 text-blue-500" />,
      lightBgColor: 'bg-gray-50',
      darkBgColor: 'dark:bg-[#4a5463]',
    },
    {
      value: '20+',
      label: 'New Sales',
      icon: <IoGameControllerSharp className="h-6 w-6 text-teal-500" />,
      lightBgColor: 'bg-[#ceeeec]',
      darkBgColor: 'dark:bg-[#366263]',
    },
    {
      value: '190+',
      label: 'New Leads',
      icon: <FaShoppingBag className="h-6 w-6 text-pink-500" />,
      lightBgColor: 'bg-[#fff0f8]',
      darkBgColor: 'dark:bg-[#4a455c]',
    },
    {
      value: '12+',
      label: 'Referral',
      icon: <MdWork className="h-6 w-6 text-blue-500" />,
      lightBgColor: 'bg-gray-50',
      darkBgColor: 'dark:bg-[#4a5463]',
    },
  ];

  return (
    <div className="grid grid-cols-1 sm:grid-cols-2 md:grid-cols-3 lg:grid-cols-4 gap-6">
      {statisticData.map((stat, index) => (
        <StatisticCard
          key={index}
          statistic={stat.value}
          label={stat.label}
          icon={stat.icon}
          lightBgColor={stat.lightBgColor}
          darkBgColor={stat.darkBgColor}
        />
      ))}
    </div>
  );
};

export default StatisticSection;
