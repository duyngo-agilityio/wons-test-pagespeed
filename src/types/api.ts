export enum Method {
  Get = 'GET',
  Post = 'POST',
  Put = 'PUT',
  Patch = 'PATCH',
  Delete = 'DELETE',
}
