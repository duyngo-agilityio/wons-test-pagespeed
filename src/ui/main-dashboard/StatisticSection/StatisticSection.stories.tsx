// Libs
import type { Meta, StoryObj } from '@storybook/react';

// components
import StatisticSection from './index';

const meta: Meta<typeof StatisticSection> = {
  title: 'Components/StatisticSection',
  component: StatisticSection,
  argTypes: {},
};

export default meta;
type Story = StoryObj<typeof StatisticSection>;

export const Interactive: Story = {
  args: {},
};
