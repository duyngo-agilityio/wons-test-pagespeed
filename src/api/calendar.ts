// Constants
import { API_PATH } from '@/constants';

// Models
import { ICalendarTask, IEvent } from '@/models';

// Services
import { httpClient } from '@/services';

// Types
import { StrapiModel, StrapiResponse } from '@/types';

// Utils
import { formatErrorMessage } from '@/utils';

export const getCalendarEvents = async (): Promise<{
  error?: string;
  data?: StrapiModel<IEvent>[];
}> => {
  try {
    const productsResponse = await httpClient.getRequest<
      StrapiResponse<StrapiModel<IEvent>[]>
    >({
      endpoint: `${API_PATH.EVENTS}?populate=users_permissions_users`,
      configOptions: {
        next: {
          tags: [API_PATH.EVENTS],
        },
      },
    });

    return { data: productsResponse.data ?? [] };
  } catch (error) {
    const message = formatErrorMessage(error);
    return { error: message };
  }
};

export const getCalendarTasks = async (): Promise<{
  error?: string;
  data?: StrapiModel<ICalendarTask>[];
}> => {
  try {
    const productsResponse = await httpClient.getRequest<
      StrapiResponse<StrapiModel<ICalendarTask>[]>
    >({
      endpoint: `${API_PATH.CALENDAR_TASKS}`,
      configOptions: {
        next: {
          tags: [API_PATH.EVENTS],
        },
      },
    });

    if (!productsResponse?.data?.length) {
      return { data: [] };
    }

    return { data: productsResponse.data };
  } catch (error) {
    const message = formatErrorMessage(error);
    return { error: message };
  }
};
