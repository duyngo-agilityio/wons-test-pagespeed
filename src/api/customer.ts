// Constants
import { API_PATH } from '@/constants';
import { ICustomer } from '@/models';

// Services
import { httpClient } from '@/services';
import { StrapiModel, StrapiResponse } from '@/types';

// Utils
import { formatErrorMessage } from '@/utils';

export const getCustomers = async (): Promise<{
  error?: string;
  data?: StrapiModel<ICustomer>[];
}> => {
  try {
    const productsResponse = await httpClient.getRequest<
      StrapiResponse<StrapiModel<ICustomer>[]>
    >({
      endpoint: API_PATH.CUSTOMERS,
    });

    if (!productsResponse?.data?.length) {
      return { error: undefined, data: [] };
    }

    return { error: undefined, data: productsResponse.data };
  } catch (error) {
    const message = formatErrorMessage(error);
    return { error: message };
  }
};
