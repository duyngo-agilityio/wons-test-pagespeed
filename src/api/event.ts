// Constants
import { API_PATH } from '@/constants';

// Models
import { IEvent } from '@/models';

// Services
import { httpClient } from '@/services';

// Types
import { StrapiModel, StrapiResponse } from '@/types';

// Utils
import { formatErrorMessage } from '@/utils';

export const getEvents = async (): Promise<{
  error?: string;
  data?: StrapiModel<IEvent>[];
}> => {
  try {
    const productsResponse = await httpClient.getRequest<
      StrapiResponse<StrapiModel<IEvent>[]>
    >({
      endpoint: API_PATH.EVENTS,
      configOptions: {
        next: {
          tags: [API_PATH.EVENTS],
        },
      },
    });

    if (!productsResponse?.data?.length) {
      return { error: undefined, data: [] };
    }

    return { error: undefined, data: productsResponse.data };
  } catch (error) {
    const message = formatErrorMessage(error);
    return { error: message };
  }
};
