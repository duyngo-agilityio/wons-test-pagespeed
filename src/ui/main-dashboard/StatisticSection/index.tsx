// Icons
import { FaHeart, FaShoppingBag } from 'react-icons/fa';
import { IoGameControllerSharp } from 'react-icons/io5';
import { MdWork } from 'react-icons/md';

// Components
import { StatisticCard, NoProductAvailable } from '@/components';

// apis
import { getAllStatistics } from '@/api';

const StatisticSection = async () => {
  const { data: statisticsData } = await getAllStatistics();

  return (
    <div className="flex flex-col lg:flex-row">
      {statisticsData?.length ? (
        statisticsData
          .map((stat) => [
            {
              value: `${stat.favoriteServices}+`,
              label: 'Favorite',
              icon: <FaHeart className="h-5 w-5 text-[#605cf8]" />,
              lightBgColor: 'bg-gray-50',
              darkBgColor: 'dark:bg-[#4a5463]',
            },
            {
              value: `${stat.newSales}+`,
              label: 'New Sales',
              icon: <IoGameControllerSharp className="h-5 w-5 text-teal-500" />,
              lightBgColor: 'bg-[#ceeeec]',
              darkBgColor: 'dark:bg-[#366263]',
            },
            {
              value: `${stat.newLeads}+`,
              label: 'New Leads',
              icon: <FaShoppingBag className="h-5 w-5 text-pink-500" />,
              lightBgColor: 'bg-[#fff0f8]',
              darkBgColor: 'dark:bg-[#4a455c]',
            },
            {
              value: `${stat.referrals}+`,
              label: 'Referral',
              icon: <MdWork className="h-5 w-5 text-[#605cf8]" />,
              lightBgColor: 'bg-gray-50',
              darkBgColor: 'dark:bg-[#4a5463]',
            },
          ])
          .flat()
          .map((stat, index) => {
            const hasBorder = index < statisticsData.length - 1;
            return (
              <div key={index} className="w-full">
                <StatisticCard
                  statistic={stat.value}
                  label={stat.label}
                  icon={stat.icon}
                  lightBgColor={stat.lightBgColor}
                  darkBgColor={stat.darkBgColor}
                />
                {hasBorder && (
                  <div className="border-t border-gray-300 dark:border-gray-600 my-8" />
                )}
              </div>
            );
          })
      ) : (
        <NoProductAvailable message="No statistics available at the moment" />
      )}
    </div>
  );
};

export default StatisticSection;
