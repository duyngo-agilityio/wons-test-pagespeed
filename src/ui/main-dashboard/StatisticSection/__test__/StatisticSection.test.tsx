import { FunctionComponent } from 'react';
import { render, waitFor } from '@testing-library/react';

// api
import { getAllStatistics } from '@/api';

// components
import StatisticSection from '../index';

jest.mock('@/api', () => ({
  ...jest.requireActual('@/api'),
  getAllStatistics: jest.fn(),
}));

async function resolvedComponent(
  Component: FunctionComponent<unknown>,
  props: unknown,
) {
  const ComponentResolved = await Component(props);
  return () => ComponentResolved;
}

describe('StatisticSection component', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render all content with resolved data and match snapshot', async () => {
    (getAllStatistics as jest.Mock).mockResolvedValue({
      data: [
        {
          favoriteServices: 5,
          newSales: 10,
          newLeads: 15,
          referrals: 20,
        },
      ],
    });

    const StatisticSectionResolved = await resolvedComponent(
      StatisticSection,
      {},
    );
    const { container } = render(<StatisticSectionResolved />);

    await waitFor(() => {
      expect(container).toMatchSnapshot();
    });
  });

  it('should render NoProductAvailable when no data is available and match snapshot', async () => {
    (getAllStatistics as jest.Mock).mockResolvedValue({
      data: [],
    });

    const StatisticSectionResolved = await resolvedComponent(
      StatisticSection,
      {},
    );
    const { container } = render(<StatisticSectionResolved />);

    await waitFor(() => {
      expect(container).toMatchSnapshot();
    });
  });
});
