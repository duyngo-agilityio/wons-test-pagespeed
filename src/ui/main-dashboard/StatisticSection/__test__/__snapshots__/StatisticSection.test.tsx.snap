// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatisticSection component should render NoProductAvailable when no data is available and match snapshot 1`] = `
<div>
  <div
    class="grid grid-cols-1 lg:grid-cols-2 xl:grid-cols-4 gap-8"
  >
    <div
      class="text-center text-gray-500 dark:text-gray-400 "
    >
      No statistics available at the moment
    </div>
  </div>
</div>
`;

exports[`StatisticSection component should render all content with resolved data and match snapshot 1`] = `
<div>
  <div
    class="grid grid-cols-1 lg:grid-cols-2 xl:grid-cols-4 gap-8"
  >
    <div
      class="w-full"
    >
      <div
        class="flex flex-col relative overflow-hidden h-auto text-foreground box-border outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 shadow-medium rounded-large transition-transform-background motion-reduce:transition-none p-2 sm:p-4 xl:p-8 2xl:py-10 2xl:px-8 bg-white rounded-10 dark:bg-gray-400"
        tabindex="-1"
      >
        <div
          class="flex items-center flex-wrap"
        >
          <div
            class="mr-2 sm:mr-4 rounded-full flex items-center justify-center w-[50px] sm:w-[63px] h-[50px] sm:h-[63px] min-w-[40px] sm:min-w-[50px] bg-gray-50 dark:bg-[#4a5463]"
            style="width: 60px; height: 60px; min-width: 45px;"
          >
            <svg
              class="h-5 w-5 text-[#605cf8]"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M462.3 62.6C407.5 15.9 326 24.3 275.7 76.2L256 96.5l-19.7-20.3C186.1 24.3 104.5 15.9 49.7 62.6c-62.8 53.6-66.1 149.8-9.9 207.9l193.5 199.8c12.5 12.9 32.8 12.9 45.3 0l193.5-199.8c56.3-58.1 53-154.3-9.8-207.9z"
              />
            </svg>
          </div>
          <div
            class="flex-1 min-w-0"
          >
            <p
              class="font-dm-sans text-blue-800 dark:text-white text-md font-bold text-[14px] sm:text-[18px] lg:text-[16px] xl:text-[24.06px] !text-blue-400 dark:!text-gray-850 leading-[31.33px]"
            >
              5+
            </p>
            <p
              class="font-dm-sans text-blue-800 dark:text-white text-md !text-blue-400 dark:!text-gray-850 w-full truncate overflow-hidden whitespace-nowrap !text-[15.02px] leading-[19.56px]"
            >
              Favorite
            </p>
          </div>
        </div>
      </div>
    </div>
    <div
      class="w-full"
    >
      <div
        class="flex flex-col relative overflow-hidden h-auto text-foreground box-border outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 shadow-medium rounded-large transition-transform-background motion-reduce:transition-none p-2 sm:p-4 xl:p-8 2xl:py-10 2xl:px-8 bg-white rounded-10 dark:bg-gray-400"
        tabindex="-1"
      >
        <div
          class="flex items-center flex-wrap"
        >
          <div
            class="mr-2 sm:mr-4 rounded-full flex items-center justify-center w-[50px] sm:w-[63px] h-[50px] sm:h-[63px] min-w-[40px] sm:min-w-[50px] bg-[#ceeeec] dark:bg-[#366263]"
            style="width: 60px; height: 60px; min-width: 45px;"
          >
            <svg
              class="h-5 w-5 text-teal-500"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M478.07 356.88 439 151c-8.86-40.35-23-71-88-71H145c-66 0-79.14 30.65-88 71L18 356.88c-7 34.12 4.43 61.25 33.37 71.81S103 423 119.18 391.3l15.42-30.52a16 16 0 0 1 14.28-8.78h198.28a16 16 0 0 1 14.28 8.78l15.42 30.52c16.14 31.7 38.88 48 67.81 37.39S485 391 478.07 356.88zM224 240h-48v48h-32v-48H96v-32h48v-48h32v48h48zm68 4a20 20 0 1 1 20-20 20 20 0 0 1-20 20zm44 44a20 20 0 1 1 20-20 20 20 0 0 1-20 20zm0-88a20 20 0 1 1 20-20 20 20 0 0 1-20 20zm44 44a20 20 0 1 1 20-20 20 20 0 0 1-20 20z"
              />
            </svg>
          </div>
          <div
            class="flex-1 min-w-0"
          >
            <p
              class="font-dm-sans text-blue-800 dark:text-white text-md font-bold text-[14px] sm:text-[18px] lg:text-[16px] xl:text-[24.06px] !text-blue-400 dark:!text-gray-850 leading-[31.33px]"
            >
              10+
            </p>
            <p
              class="font-dm-sans text-blue-800 dark:text-white text-md !text-blue-400 dark:!text-gray-850 w-full truncate overflow-hidden whitespace-nowrap !text-[15.02px] leading-[19.56px]"
            >
              New Sales
            </p>
          </div>
        </div>
      </div>
    </div>
    <div
      class="w-full"
    >
      <div
        class="flex flex-col relative overflow-hidden h-auto text-foreground box-border outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 shadow-medium rounded-large transition-transform-background motion-reduce:transition-none p-2 sm:p-4 xl:p-8 2xl:py-10 2xl:px-8 bg-white rounded-10 dark:bg-gray-400"
        tabindex="-1"
      >
        <div
          class="flex items-center flex-wrap"
        >
          <div
            class="mr-2 sm:mr-4 rounded-full flex items-center justify-center w-[50px] sm:w-[63px] h-[50px] sm:h-[63px] min-w-[40px] sm:min-w-[50px] bg-[#fff0f8] dark:bg-[#4a455c]"
            style="width: 60px; height: 60px; min-width: 45px;"
          >
            <svg
              class="h-5 w-5 text-pink-500"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 448 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M352 160v-32C352 57.42 294.579 0 224 0 153.42 0 96 57.42 96 128v32H0v272c0 44.183 35.817 80 80 80h288c44.183 0 80-35.817 80-80V160h-96zm-192-32c0-35.29 28.71-64 64-64s64 28.71 64 64v32H160v-32zm160 120c-13.255 0-24-10.745-24-24s10.745-24 24-24 24 10.745 24 24-10.745 24-24 24zm-192 0c-13.255 0-24-10.745-24-24s10.745-24 24-24 24 10.745 24 24-10.745 24-24 24z"
              />
            </svg>
          </div>
          <div
            class="flex-1 min-w-0"
          >
            <p
              class="font-dm-sans text-blue-800 dark:text-white text-md font-bold text-[14px] sm:text-[18px] lg:text-[16px] xl:text-[24.06px] !text-blue-400 dark:!text-gray-850 leading-[31.33px]"
            >
              15+
            </p>
            <p
              class="font-dm-sans text-blue-800 dark:text-white text-md !text-blue-400 dark:!text-gray-850 w-full truncate overflow-hidden whitespace-nowrap !text-[15.02px] leading-[19.56px]"
            >
              New Leads
            </p>
          </div>
        </div>
      </div>
    </div>
    <div
      class="w-full"
    >
      <div
        class="flex flex-col relative overflow-hidden h-auto text-foreground box-border outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 shadow-medium rounded-large transition-transform-background motion-reduce:transition-none p-2 sm:p-4 xl:p-8 2xl:py-10 2xl:px-8 bg-white rounded-10 dark:bg-gray-400"
        tabindex="-1"
      >
        <div
          class="flex items-center flex-wrap"
        >
          <div
            class="mr-2 sm:mr-4 rounded-full flex items-center justify-center w-[50px] sm:w-[63px] h-[50px] sm:h-[63px] min-w-[40px] sm:min-w-[50px] bg-gray-50 dark:bg-[#4a5463]"
            style="width: 60px; height: 60px; min-width: 45px;"
          >
            <svg
              class="h-5 w-5 text-[#605cf8]"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="M20 6h-4V4c0-1.11-.89-2-2-2h-4c-1.11 0-2 .89-2 2v2H4c-1.11 0-1.99.89-1.99 2L2 19c0 1.11.89 2 2 2h16c1.11 0 2-.89 2-2V8c0-1.11-.89-2-2-2zm-6 0h-4V4h4v2z"
              />
            </svg>
          </div>
          <div
            class="flex-1 min-w-0"
          >
            <p
              class="font-dm-sans text-blue-800 dark:text-white text-md font-bold text-[14px] sm:text-[18px] lg:text-[16px] xl:text-[24.06px] !text-blue-400 dark:!text-gray-850 leading-[31.33px]"
            >
              20+
            </p>
            <p
              class="font-dm-sans text-blue-800 dark:text-white text-md !text-blue-400 dark:!text-gray-850 w-full truncate overflow-hidden whitespace-nowrap !text-[15.02px] leading-[19.56px]"
            >
              Referral
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
